### PR TITLE
Support missed Object(Nullable(json)) subcolumns in query.

### LINKED ISSUE #40990: Delete rows bug for JSON column type
clickhouse version 22.8
doesn't delete row from table using alter table delete command

create table query:
```
CREATE TABLE logs
(
    datetime DEFAULT now() Codec(DoubleDelta, LZ4),
    message JSON
)
ENGINE = MergeTree
order by datetime;
```

insert query:
```
INSERT INTO logs(message) VALUES ('{"a": 2, "b": { "c": 22, "d": [12, 122, 13] }, "f": "a"}');
```

delete query:
```
alter table logs delete where message.f = 'a';
```

error from select * from system.mutations;
```
Code: 47. DB::Exception: Missing columns: 'message.f' while processing query: 'SELECT count() WHERE (message.f = 'a') OR (message.f = 'a')', required columns: 'message.f', maybe you meant: ['message']. (UNKNOWN_IDENTIFIER) (version 22.3.8.40.altinitystable (altinity build))
```

### PR DESCRIPTION
**Pull request by VanDarkholme7**
Fixes #40990  and Improves #54112 

Currently if there are missed JSON subcolumns in query, it will throw exceptions like:
`Code: 47. DB::Exception: Missing columns ...`

Since JSON type is schema free, we should better ignore missed subcolumns instead of throwing exceptions.

For example:
`CREATE TABLE test
(a UInt64, b JSON) ENGINE=MergeTree ORDER BY a;
`
Then insert some data:
`INSERT INTO test VALUES (1,'{"a":123}'), (2,'{"b":222}');  `

If we use missing subcolumns, it will throw an exception:
`SELECT a, b.a, b.b, b.c FROM test WHERE a<3 or b.c<10`

To avoid exceptions, we can implicitly rewrite the select query :
`SELECT a, b.a, b.b, NULL as b.c FROM test WHERE a<3 or (NULL as b.c)<10`

Currently it only supports `Object(Nullable('json'))` type.

### Changelog category (leave one):
- Bug Fix

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Replace missed Object(Nullable('json')) subcolumns to NULL in query instead of throwing an exception.

### CODE REVIEW

**Comment by CurtizJ** [comment]
`SELECT a, b.a, b.b, b.c FROM test WHERE a < 3 or b.c < 10` is rewritten `to SELECT a, b.a, b.b, '' as b.c FROM test WHERE a < 3 or false`.

This is inconsistent. What if there will be condition `b.c = 0` or `b.c = ""`. They should evaluate to `true`, but will evaluate to `false`. And what if there will be an expression that implies that `b.c` has integer type, e.g. `b.c * 2` - a weird exception will be thrown. I think current behaviour is more predictable and more consistent. I don't see any good solutions except replacing missed subcolumns to `NULL`. It will work as expected with `Object(Nullable('json'))`, but will be inconsistent with simple `Object('json')` because it has non-nullable subcolumns. If the development of this feature will be ever continued, the `Object` type will have nullable subcolumns as the only option, I think, because it's more correct, and replacing to `NULL` won't be a problem.

Also you have to  implement this improvement in new analyzer as well, but I'd better to close this PR.

**Review by CurtizJ — CHANGES_REQUESTED** [review]
`SELECT a, b.a, b.b, b.c FROM test WHERE a < 3 or b.c < 10` is rewritten `to SELECT a, b.a, b.b, '' as b.c FROM test WHERE a < 3 or false`.

This is inconsistent. What if there will be condition `b.c = 0` or `b.c = ""`. They should evaluate to `true`, but will evaluate to `false`. And what if there will be an expression that implies that `b.c` has integer type, e.g. `b.c * 2` - a weird exception will be thrown. I think current behaviour is more predictable and more consistent. I don't see any good solutions except replacing missed subcolumns to `NULL`. It will work as expected with `Object(Nullable('json'))`, but will be inconsistent with simple `Object('json')` because it has non-nullable subcolumns. If the development of this feature will be ever continued, the `Object` type will have nullable subcolumns as the only option, I think, because it's more correct, and replacing to `NULL` won't be a problem.

Also you have to  implement this improvement in new analyzer as well, but I'd better to close this PR.

**Comment by VanDarkholme7** [comment]
You are right, I didn't consider such cases.

**Comment by VanDarkholme7** [comment]
> I don't see any good solutions except replacing missed subcolumns to `NULL`. It will work as expected with `Object(Nullable('json'))`, but will be inconsistent with simple `Object('json')` because it has non-nullable subcolumns. 

So you think it's better to replace missed subcolumns to `NULL` in case the JSON type is `Nullable`, and when it's simple `Object('json')` type, we should throw an `Missing columns` exception?

**Comment by UnamedRus** [comment]
It would be nice to have this support also for DEFAULT/ALIAS/MATERIALIZE expressions in table definition
https://fiddle.clickhouse.com/30dde62d-294a-465c-86bd-fc1f8d14e57f
```
CREATE TABLE test
(
    `a` UInt64,
    `b` JSON,
    `c` DEFAULT b.aaa
)
ENGINE = MergeTree
ORDER BY a

```

**Comment by VanDarkholme7** [comment]
Hi @CurtizJ. According to your comment, I have rewritten this feature, now it replaces missed subcolumns to `NULL` and only supports for `Object(Nullable('json'))` type.

Commit  https://github.com/ClickHouse/ClickHouse/pull/54052/commits/cf8c614fedd7f01836a7124998ab9a393bbde8fe fixes #40990.

Commit https://github.com/ClickHouse/ClickHouse/pull/54052/commits/bfd99cde262be898dd0adb15d31209947631b158 add support for replacing missed subcolumns to `NULL`. It rewrites `AST` by replacing `ASTIdentifier` with missed json subcolumns to `ASTLiteral(NULL)` in `SELECT` clause and filtration clauses like `WHERE`, `PREWHERE` and `HAVING`.

Commit https://github.com/ClickHouse/ClickHouse/pull/54052/commits/b17736e968fb8d7114c5dca0d8d2beab1a04c86a add support for the new analyzer. It rewrites `QueryTree` by replacing `Identifier` with missed json subcolumns to `ConstantNode(NULL)`.

**Comment by VanDarkholme7** [comment]
> It would be nice to have this support also for DEFAULT/ALIAS/MATERIALIZE expressions in table definition 

Yes, it should be supported if this PR is accepted.

**Comment by VanDarkholme7** [comment]
> Commit [cf8c614](https://github.com/ClickHouse/ClickHouse/commit/cf8c614fedd7f01836a7124998ab9a393bbde8fe) fixes #40990.
> 
> Commit [78660eb](https://github.com/ClickHouse/ClickHouse/commit/78660eb859e3ac9fd1af50ff5bf00b63bb90d27e) add support for replacing missed subcolumns to `NULL`. It rewrites `AST` by replacing `ASTIdentifier` with missed json subcolumns to `ASTLiteral(NULL)`.
> 
> Commit [349b868](https://github.com/ClickHouse/ClickHouse/commit/349b868a6220a94cb5131e69c02ee7f72f22b97e) add support for the new analyzer. It rewrites `QueryTree` by replacing `Identifier` with missed json subcolumns to `ConstantNode(NULL)`.

Hi! @CurtizJ . It seems the failed `BuilderDeb` tests are not related. Could you please take a look?

**Comment by VanDarkholme7** [comment]
`Integration tests (release) [2/4] — fail: 50, passed: 548`

- `test_storage_rabbitmq/test.py`

It seems the failed integration tests are not related. Could you please take a look? @CurtizJ

**Comment by feixue1121** [comment]
@CurtizJ hi Curitiz, we expect this bug fix to support updating json column successfully even when there are no subcolumns in some parts. Could you please take a moment to review changes made by @VanDarkholme7 and point out deficiencies? Thanks.

File: src/Storages/MergeTree/StorageFromMergeTreeDataPart.h [hunk 1]
```diff
@@ -59,6 +59,21 @@ class StorageFromMergeTreeDataPart final : public IStorage
...
+        if (!hasDynamicSubcolumns(storage_columns))
+            return std::make_shared<StorageSnapshot>(*this, metadata_snapshot);
+
+        auto data_parts = storage.getDataPartsVectorForInternalUsage();
+
+        auto object_columns = getConcreteObjectColumns(
+            data_parts.begin(), data_parts.end(), storage_columns, [](const auto & part) -> const auto & { return part->getColumns(); });
+
+        return std::make_shared<StorageSnapshot>(*this, metadata_snapshot, std::move(object_columns));
+    }
```
Comment (CurtizJ) [inline src/Storages/MergeTree/StorageFromMergeTreeDataPart.h:75]: Please, leave only `getStorageSnapshot`. `getStorageSnapshotForQuery` calls `getStorageSnapshot` by default and for this storage overriding of `getStorageSnapshotForQuery` is not needed.

**Review by CurtizJ — COMMENTED** [review]
In general ok.

File: src/Storages/MergeTree/StorageFromMergeTreeDataPart.h (hunk 1 above)
Comment (VanDarkholme7) [reply src/Storages/MergeTree/StorageFromMergeTreeDataPart.h]: Fixed. https://github.com/ClickHouse/ClickHouse/pull/54052/commits/7dda1646fd8c507ee32b53e35db3d7493150bd6c

**Comment by VanDarkholme7** [comment]
All checks have passed, please take a look. @CurtizJ

File: src/Analyzer/Passes/QueryAnalysisPass.cpp [hunk 2]
```diff
@@ -3036,6 +3051,29 @@ QueryTreeNodePtr QueryAnalyzer::tryResolveIdentifierFromJoin(const IdentifierLoo
...
+        if (left_resolved_column.getValueStringRepresentation() == "NULL")
+            return left_resolved_identifier;
+    }
+    else if (right_resolved_identifier && right_resolved_identifier->getNodeType() == QueryTreeNodeType::CONSTANT)
+    {
+        resolved_side = JoinTableSide::Right;
+        auto & right_resolved_column = right_resolved_identifier->as<ConstantNode &>();
+        if (right_resolved_column.getValueStringRepresentation() == "NULL")
+            return right_resolved_identifier;
+    }
```
[the remark this answers is not in the document]
Comment (VanDarkholme7) [reply src/Analyzer/Passes/QueryAnalysisPass.cpp:3077]: @CurtizJ Added comment. https://github.com/ClickHouse/ClickHouse/pull/54052/commits/7dda1646fd8c507ee32b53e35db3d7493150bd6c

**Review by CurtizJ — APPROVED** [review]

**Pull request merged**